### PR TITLE
feat: RFC0434 out-of-band - support connection reuse

### PR DESCRIPTION
**Pull request by llorllale**
Signed-off-by: George Aristy <george.aristy@securekey.com>

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -42,4 +42,9 @@ require (
 	nhooyr.io/websocket v1.8.3
 )
 
+replace (
+	github.com/hyperledger/aries-framework-go/component/storageutil => ./component/storageutil
+	github.com/hyperledger/aries-framework-go/spi => ./spi
+)
```
Comment (llorllale) [inline go.mod:48]: Can someone please confirm the validity of this change?

Comment (Baha-sk) [reply go.mod:48]: there should be no `replace` statements for storage related packages in the root go.mod

cc @DRK3 @troyronda

Comment (DRK3) [reply go.mod]: Agreed @Baha-sk

Comment (llorllale) [reply go.mod]: Removed

**Review by Baha-sk — APPROVED** [review]

File: pkg/didcomm/protocol/introduce/service.go [hunk 2]
```diff
@@ -258,12 +257,13 @@ func (s *Service) doHandle(msg service.DIDCommMsg, outbound bool) (*metaData, er
...
 func (s *Service) OOBMessageReceived(msg service.StateMsg) error {
-	if msg.StateID != stateOOBInvited || msg.Type != service.PostState || msg.Msg.ParentThreadID() == "" {
+	// TODO we should verify if this parent thread ID is an instance of the introduce protocol
+	if msg.StateID != stateOOBInitial || msg.Type != service.PreState || msg.Msg.ParentThreadID() == "" {
 		return nil
 	}
 
 	// NOTE: the message is being used internally.
-	// Do not modify the payload such as ID and Thread.
+	// Do not modify the paylo1d such as ID and Thread.
```
Comment (troyronda) [inline pkg/didcomm/protocol/introduce/service.go:266]: typo

File: test/bdd/features/introduce.feature [hunk 3]
```diff
@@ -54,18 +54,21 @@ Feature: Introduce protocol
...
-    And   "Alice-Router" creates a route exchange client
-    When   "Alice" sends introduce proposal to the "Alice-Router" and "Bob"
-    And   "Alice-Router" wants to know "Bob" and sends introduce response with approve and provides an out-of-band invitation with an embedded "route-request"
-    And   "Bob" wants to know "Alice-Router" and sends introduce response with approve
-    Then   "Alice" checks the history of introduce protocol events "arranging,arranging,arranging,arranging,arranging,arranging,arranging,arranging,delivering,delivering,confirming,confirming,done,done"
-    And   "Alice-Router" checks the history of introduce protocol events "deciding,deciding,waiting,waiting,done,done"
-    Then  "Bob" has did exchange connection with "Alice-Router"
-    And   "Bob" checks the history of introduce protocol events "deciding,deciding,waiting,waiting,done,done" and stop
-    Then  "Bob" confirms route registration with "Alice-Router"
+    # TODO this test must be reimplemented after outofband was updated.
```
Comment (troyronda) [inline test/bdd/features/introduce.feature:57]: TODO

File: test/bdd/bddtests_test.go [hunk 4]
```diff
@@ -196,7 +195,6 @@ func features() []feature {
 		outofband.NewOutofbandControllerSteps(),
 		presentproof.NewPresentProofSDKSteps(),
 		presentproof.NewPresentProofControllerSteps(),
-		redeemableroutes.NewBDDSteps(),
```
Comment (troyronda) [inline test/bdd/bddtests_test.go:199]: what happened to this?

File: test/bdd/features/redeemable_routes.feature [hunk 5]
```diff
@@ -1,24 +0,0 @@
-#
```
Comment (troyronda) [inline test/bdd/features/redeemable_routes.feature:1]: what happened to this file?

File: test/bdd/bddtests_test.go (hunk 4 above)
Comment (llorllale) [reply test/bdd/bddtests_test.go:199]: The `introduce` BDD test [here](https://github.com/hyperledger/aries-framework-go/pull/2729#discussion_r613715014) is intended to demonstrate the same thing.

File: test/bdd/features/redeemable_routes.feature (hunk 5 above)
Comment (llorllale) [reply test/bdd/features/redeemable_routes.feature:1]: See https://github.com/hyperledger/aries-framework-go/pull/2729#discussion_r614048232

File: test/bdd/features/introduce.feature (hunk 3 above)
Comment (llorllale) [reply test/bdd/features/introduce.feature:57]: created #2735

File: pkg/didcomm/protocol/introduce/service.go (hunk 2 above)
Comment (llorllale) [reply pkg/didcomm/protocol/introduce/service.go]: fixed

**Review by fqutishat — APPROVED** [review]

**Pull request merged**